### ISSUE TITLE
marshmallow-sqlalchemy support

### ISSUE THREAD

**Issue by zikphil** [jmcarp/flask-apispec#73]
I am encountering issues when trying to use the `use_kwargs` decorator on Schemas that have been generated from SQLAlchemy models using the marshmallow-sqlalchemy package.

```from marshmallow_sqlalchemy import ModelSchema

class Store(db.Model):
    __tablename__ = 'stores'

    strict = True
    id = Column(Integer, primary_key=True, autoincrement=True)
    created = Column(DateTime, default=datetime.datetime.now)
    updated = Column(DateTime, default=datetime.datetime.now)
    active = Column(Boolean, default=True)
    owner = Column(String)
    name = Column(String)


class StoreSchema(ModelSchema):
    class Meta:
        model = Store
        sqla_session = db.session
        strict = True
```

```
from flask import make_response, jsonify, abort
from flask_apispec import MethodResource
from flask_apispec import use_kwargs, marshal_with, doc
from .models import StoreSchema, Store, store_schema
from . import db


@doc(description='Store Information', tags=['store'])
@marshal_with(StoreSchema)
class StoreResource(MethodResource):

    def get(self, store_id):
        store = Store.query.get_or_404(store_id)
        return store_schema.dump(store).data

    @use_kwargs(StoreSchema)
    def put(self, store_id, **kwargs):
        store = Store.query.get_or_404(store_id)
        store.__dict__.update(**kwargs)
        db.session.commit()
        return store_schema.dump(store).data
```

```Traceback (most recent call last):
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/werkzeug/serving.py", line 209, in run_wsgi
    execute(self.server.app)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/werkzeug/serving.py", line 197, in execute
    application_iter = app(environ, start_response)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1985, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask_restful/__init__.py", line 273, in error_router
    return original_handler(e)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1540, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/_compat.py", line 32, in reraise
    raise value.with_traceback(tb)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask_restful/__init__.py", line 273, in error_router
    return original_handler(e)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/_compat.py", line 32, in reraise
    raise value.with_traceback(tb)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask_restful/__init__.py", line 480, in wrapper
    resp = resource(*args, **kwargs)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/views.py", line 84, in view
    return self.dispatch_request(*args, **kwargs)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask/views.py", line 149, in dispatch_request
    return meth(*args, **kwargs)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask_apispec/annotations.py", line 117, in wrapped
    return wrapper(*args, **kwargs)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask_apispec/wrapper.py", line 23, in __call__
    response = self.call_view(*args, **kwargs)
  File "/Users/phil/Library/Python/3.6/lib/python/site-packages/flask_apispec/wrapper.py", line 41, in call_view
    kwargs.update(parsed)
TypeError: 'Store' object is not iterable
```

Does it make sense? I wonder if this is an implementation issue or just an incompatibility.

**Comment by zikphil** [comment]
Wierd fix, not sure if it's the root cause but changing Line 39 of wrapper.py to 

`parsed = parser.parse(schema.fields, locations=option['kwargs']['locations'])`

from 

`parsed = parser.parse(schema, locations=option['kwargs']['locations'])`

Fixed the issue for me. Any insights to give me on why this should not be for everyone?

**Comment by jjanyan** [comment]
I have the same issue. Changing to schema.fields does the trick. 

@zikphil what did you end up doing for your fix?

**Comment by jjanyan** [comment]
I did some more digging. 

`schema.fields` is a dictionary defining the fields and how to validate them according to marshmallows rules.

`schema` is similar to` schema.fields`, but it indicates that an object should be created not a dictionary. As defined by `Schema.Meta.model`. 

If you pass in `schema.fields`, webargs generates a validated dictionary, which can be iterated over to update kwargs (line 41 in wrapper.py)

If you pass in `schema`, webargs generates an object, which can NOT be iterated over to update kwargs. 

Knowing that, we just need to force apispec and webargs to use fields and not the model-based schema. The simple/easy solution was to send in the marshmallow schema fields to flask_apispec.use_kwargs.

```
    @use_kwargs(StoreSchema().fields)
    def put(self, store_id, **kwargs):
        store = Store.query.get_or_404(store_id)
        store.__dict__.update(**kwargs)
        db.session.commit()
        return store_schema.dump(store).data
```

TLDR; use `StoreSchema().fields` instead of `StoreSchema`

**Comment by livingmine** [comment]
@jjanyan I encountered the same issue and followed your suggestion. While it solved the issue, it seems that the validation process defined in the schema is somehow bypassed.
```

class StoreSchema(ModelSchema):
    class Meta:
        model = Store
        sqla_session = db.session
        strict = True
    
    @validates('active')
    def validate_active(self, value):
        ...

```

**Comment by jjanyan** [comment]
@livingmine yes, it will ignore any custom validation. Schema.fields() returns the marshmallow fields, as if you had passed in a dict like `{"name":ma.fields.Str()}` (don't quote me on the exact format).

Ideally we would add an extra check in wrapper.py for marshmallow-sqlachemy models, in which case your `validate_active` method would execute. 

There doesn't seem to be much activity since October, so I worked around it instead of waiting.

**Comment by livingmine** [comment]
Okay, i ended up with another solution. I'm still using the original StoreSchema instead of StoreSchema().fields to preserve the validation processes. So instead, i monkey patched the wrapper.py by adding
```
if 'make_instance' in dir(schema):
    parsed = schema.dump(parsed).data
```
under 
```
parsed = parser.parse(schema, locations=option['kwargs']['locations'])
```
to convert the returned instance/object to dictionary.

**Comment by jkida** [comment]
Why not just override make_instance to return a dict?
```
class UserSchema(ma.ModelSchema):
    class Meta:
        model = User
        strict = True

    @marshmallow.post_load
    def make_instance(self, data):
        return data
```

My Resource is just defined with:
```
@api.route('/user')
class UsersAPI(Resource):
    @use_kwargs(UserSchema)
    @marshal_with(UserSchema
    def post(self, **data):
        user = User(**data)
        #add and commit
        return user
```

Referencing: https://github.com/sloria/webargs/issues/126
